### PR TITLE
Add proof generation and verification

### PR DESCRIPTION
**Pull request by AurelienFT**
This PR adds the possibility to generate proof that a key/value is or isn't in the trie. The proof is a list of node with their hash and the ones of their children following the same format as Madara.

You can also verify the proof using a new function that will verify if the proof is indeed correct and that it verify if the key/value is (or isn't) in the trie.

I have updated the example in the README.md for usage.

### CODE REVIEW

File: src/trie/merkle_tree.rs [hunk 1]
```diff
@@ -112,6 +118,39 @@ fn test_shared_path_encode_decode() {
     let decoded = Path::decode(&mut &encoded[..]).unwrap();
     assert_eq!(path, decoded);
 }
+/// A node used in proof generated by the trie.
+///
+/// See pathfinders merkle-tree crate for more information.
```
Comment (0xLucqs) [inline src/trie/merkle_tree.rs]: can you either add the link to the doc or add a small summary plz

File: src/trie/merkle_tree.rs [hunk 2]
```diff
@@ -112,6 +118,39 @@ fn test_shared_path_encode_decode() {
...
+}
+
+impl ProofNode {
+    pub fn hash<H: HasherT>(&self) -> Felt252Wrapper {
+        match self {
+            ProofNode::Binary { left, right } => Felt252Wrapper(H::hash_elements(left.0, right.0)),
+            ProofNode::Edge { child, path } => {
+                let mut bytes = [0u8; 32];
+                bytes.view_bits_mut::<Msb0>()[256 - path.0.len()..].copy_from_bitslice(&path.0);
+                let path_hash = Felt252Wrapper::try_from(&bytes).unwrap();
```
Comment (0xLucqs) [inline src/trie/merkle_tree.rs:144]: can you add explain why it's safe to unwrap

File: src/trie/merkle_tree.rs [hunk 3]
```diff
@@ -112,6 +118,39 @@ fn test_shared_path_encode_decode() {
...
+                let mut length = [0; 32];
+                // Safe as len() is guaranteed to be <= 251
+                length[31] = path.0.len() as u8;
+
+                let length = Felt252Wrapper::try_from(&length).unwrap();
```
Comment (0xLucqs) [inline src/trie/merkle_tree.rs]:
```suggestion
                let length = Felt252Wrapper::from(path.0.len() as u8);
```

File: src/trie/merkle_tree.rs [hunk 4]
```diff
@@ -704,9 +740,162 @@ impl<H: HasherT, DB: BonsaiDatabase, ID: Id> MerkleTree<H, DB, ID> {
...
+    ///   3. the root hash matches the known root
     ///
+    /// # Arguments
+    ///
+    /// * `key` - The key to get the merkle proof of.
+    ///
+    /// # Returns
+    ///
+    /// The merkle proof and all the child nodes hashes.
+    pub fn get_proof(&self, key: &BitSlice<u8, Msb0>) -> Result<Vec<ProofNode>, BonsaiStorageError>
```
Comment (0xLucqs) [inline src/trie/merkle_tree.rs:759]: can we make that generic over the hash function ?

File: src/trie/merkle_tree.rs [hunk 5]
```diff
@@ -891,6 +1080,86 @@ impl<H: HasherT, DB: BonsaiDatabase, ID: Id> MerkleTree<H, DB, ID> {
...
+    ///    2. move towards the target - if current is:
+    ///       1. binary node then choose the child that moves towards the target, else if
+    ///       2. edge node then check the path against the target bits
+    ///          1. If it matches then proceed with the child, else
+    ///          2. if it does not match then we now have a proof that the target does not exist
+    ///    3. nibble off target bits according to which child you got in (2). If all bits are gone then you
+    ///       have reached the target and the child hash is the value you wanted and the proof is complete.
+    ///    4. set expected_hash <- to the child hash
+    /// 3. check that the expected_hash is `value` (we should've reached the leaf)
+    pub fn verify_proof(
```
Comment (0xLucqs) [inline src/trie/merkle_tree.rs:1099]: can we make that generic over the hash function ?

**Review by 0xLucqs — CHANGES_REQUESTED** [review]

File: src/trie/merkle_tree.rs (hunk 5 above)
Comment (AurelienFT) [reply src/trie/merkle_tree.rs:1099]: Done

File: src/trie/merkle_tree.rs (hunk 4 above)
Comment (AurelienFT) [reply src/trie/merkle_tree.rs:759]: Can you point me to the code that is "hash-specific" in this function ?

File: src/trie/merkle_tree.rs (hunk 3 above)
Comment (AurelienFT) [reply src/trie/merkle_tree.rs]: Done

File: src/trie/merkle_tree.rs (hunk 2 above)
Comment (AurelienFT) [reply src/trie/merkle_tree.rs:144]: Done

File: src/trie/merkle_tree.rs (hunk 1 above)
Comment (AurelienFT) [reply src/trie/merkle_tree.rs]: Done

File: src/trie/merkle_tree.rs (hunk 2 above)
Comment (0xLucqs) [reply src/trie/merkle_tree.rs:144]: Felt is slightly less than 32 bytes it's 252 bits so 32 bytes isn't always safe

File: src/trie/merkle_tree.rs (hunk 4 above)
Comment (0xLucqs) [reply src/trie/merkle_tree.rs:759]: ah indeed all good here

**Review by 0xLucqs — CHANGES_REQUESTED** [review]

File: src/trie/merkle_tree.rs (hunk 2 above)
Comment (0xLucqs) [reply src/trie/merkle_tree.rs:144]: AFAIU it's safe because the path is 251 bits which is < than our prime

Comment (AurelienFT) [reply src/trie/merkle_tree.rs:144]: Yes sorry I mixed up multiple things. Corrected.

**Review by 0xLucqs — APPROVED** [review]
Lgtm

**Pull request merged**